core/kazoo_number_manager/src/knm_numbers.erl

```diff
@@ -637,6 +637,53 @@ maybe_set_ported_in(T=#{todo := PNs, options := Options}) ->
             knm_phone_number:setters(T, Routines)
     end.
 
+%% @private
+%% Generates options for number creation, can return error if the correct options
+%% are not passed
+-spec options_for_create(ne_binaries(), knm_number_options:options()) ->
+                                knm_number_options:options() |
+                                ret().
+options_for_create(Nums, Options0) ->
+    Fs = [fun state_create_option/2
+         ,fun module_name_create_option/2
+         ],
+    lists:foldl(fun(_F, #{}=Ret) -> Ret;
+                   (F, AccOptions) -> F(Nums, AccOptions)
+                end
+               ,Options0
+               ,Fs
+               ).
+
+%% @private
+-spec state_create_option(ne_binaries(), knm_number_options:options()) ->
+                                 knm_number_options:options() |
+                                 ret().
+state_create_option(Nums, Options) ->
+    case knm_number_options:assign_to(Options) of
+        ?MATCH_ACCOUNT_RAW(AccountId) ->
+            ToState = knm_number:state_for_create(AccountId, Options),
+            lager:debug("picked state ~s for ~s for ~p", [ToState, AccountId, Nums]),
+            [{'state', ToState} | Options];
+        _ ->
+            Error = knm_errors:to_json(assign_failure, undefined, field_undefined),
+            ret(new(Options, [], Nums, Error))
```

Comment:
I still think this test should happen at the very beginning. Also this should be using T1 instead of new.